GitUI/Avatars/AvatarDownloader.cs

```diff
@@ -61,173 +35,76 @@ public async Task<Image> GetAvatarAsync(string email, string name, int imageSize
                 return null;
             }
 
-            var match = _gitHubEmailRegex.Match(email);
-
-            return match.Success
-                ? await LoadFromGitHubAsync()
-                : await LoadFromGravatarAsync();
-
-            async Task<Image> LoadFromGitHubAsync()
-            {
-                try
-                {
-                    var username = match.Groups["username"].Value;
-                    var client = new Git.hub.Client();
-                    var user = await client.GetUserAsync(username);
-                    if (!string.IsNullOrEmpty(user?.AvatarUrl))
-                    {
-                        var builder = new UriBuilder(user.AvatarUrl);
-                        var query = new StringBuilder(builder.Query.TrimStart('?'));
-                        query.Append(query.Length == 0 ? "?" : "&");
-                        query.Append("s=").Append(imageSize);
-                        builder.Query = query.ToString();
-                        var imageUrl = builder.Uri;
-
-                        return await DownloadImageAsync(imageUrl);
-                    }
-                }
-                catch (Exception ex)
-                {
-                    Trace.WriteLine(ex.Message);
-                }
+            ClearOldCacheEntries();
 
-                return null;
-            }
+            var errorCount = 0;
 
-            Task<Image> LoadFromGravatarAsync()
+            while (true)
             {
-                return DownloadImageAsync(BuildGravatarUrl());
+                var (_, task) = _downloads.GetOrAdd(imageUrl, _ => (DateTime.UtcNow, DownloadAsync(imageUrl)));
 
-                Uri BuildGravatarUrl()
+                // If we discover a faulted task, remove it and try again
+                if (task.IsFaulted || task.IsCanceled)
                 {
-                    var d = GetDefaultImageString();
-                    var hash = CalculateHash();
+                    _downloads.TryRemove(imageUrl, out _);
```

Comment:
Why is a dictionary needed for this?

Reply:
It's basically a "copy-pasta" from the [original implementation](https://github.com/gitextensions/gitextensions/blob/8d3bc01349c19674c54beb6741cfd8224c0479e4/GitUI/Avatars/AvatarDownloader.cs#L152) and I kept it as is, because it's field tested ^^. The dictionary is used to combine simultaneous requests to the same image resource and only make a network call once (I guess)